src/GafferOSLUI/OSLImageUI.cpp

```diff
@@ -0,0 +1,326 @@
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+
+#include "GafferOSL/OSLImage.h"
+#include "GafferOSL/ClosurePlug.h"
+
+#include "GafferUI/Nodule.h"
+#include "GafferUI/NoduleLayout.h"
+#include "GafferUI/PlugAdder.h"
+
+#include "Gaffer/Metadata.h"
+#include "Gaffer/MetadataAlgo.h"
+#include "Gaffer/ScriptNode.h"
+#include "Gaffer/UndoScope.h"
+#include "Gaffer/NameValuePlug.h"
+#include "Gaffer/PlugAlgo.h"
+
+#include "IECore/CompoundData.h"
+
+#include "boost/algorithm/string.hpp"
+#include "boost/bind.hpp"
+
+using namespace std;
+using namespace Gaffer;
+using namespace GafferUI;
+
+namespace
+{
+
+std::string cleanupChannelName( std::string s )
+{
+	boost::erase_all( s, "RGBA" );
+	boost::erase_all( s, "RGB" );
+	return s;
+}
+
+class OSLImagePlugAdder : public PlugAdder
+{
+
+	public :
+
+		OSLImagePlugAdder( GraphComponentPtr plugsParent )
+			:   m_plugsParent( IECore::runTimeCast<Plug>( plugsParent ) )
+		{
+			if( ! m_plugsParent )
+			{
+				throw IECore::Exception( "OSLImageUI::PlugAdder constructor must be passed plug" );
+			}
+			buttonReleaseSignal().connect( boost::bind( &OSLImagePlugAdder::buttonRelease, this, ::_2 ) );
+		}
+
+	protected :
+
+		bool canCreateConnection( const Plug *endpoint ) const override
+		{
+			IECore::ConstCompoundDataPtr plugAdderOptions = Metadata::value<IECore::CompoundData>( m_plugsParent->node(), "plugAdderOptions" );
+			return !availableChannels( plugAdderOptions.get(), endpoint ).empty();
+		}
+
+		void createConnection( Plug *endpoint ) override
+		{
+			IECore::ConstCompoundDataPtr plugAdderOptions = Metadata::value<IECore::CompoundData>( m_plugsParent->node(), "plugAdderOptions" );
+			vector<std::string> names = availableChannels( plugAdderOptions.get(), endpoint );
+
+			std::string picked = menuSignal()( "Connect To", names );
+			if( !picked.size() )
+			{
+				return;
+			}
+
+			NameValuePlug *newPlug = addPlug( cleanupChannelName( picked ), plugAdderOptions->member<IECore::Data>( picked ) );
+			newPlug->valuePlug()->setInput( endpoint );
+		}
+
+	private :
+		std::set<std::string> usedNames() const
+		{
+			std::set<std::string> used;
+			for( NameValuePlugIterator it( m_plugsParent.get() ); !it.done(); ++it )
+			{
+				// TODO - this method for checking if a plug variesWithContext should probably live in PlugAlgo
+				// ( it's based on Switch::variesWithContext )
+				PlugPtr sourcePlug = (*it)->namePlug()->source<Gaffer::Plug>();
+				bool variesWithContext = sourcePlug->direction() == Plug::Out && IECore::runTimeCast<const ComputeNode>( sourcePlug->node() );
+				if( !variesWithContext )
+				{
+					used.insert( (*it)->namePlug()->getValue() );
+				}
+			}
+			return used;
+		}
+
+		NameValuePlug* addPlug( std::string channelName, const IECore::Data *defaultData )
```

Comment:
I hadn't realised quite how much code we'd be duplicating when I advocated for not having a python PlugAdder. At some point please badger me to review and/or improve the performance for those (and the `typeId()` override stuff we discussed the other day) so we can consolidate everything.